include/deal.II/base/array_view.h

```diff
@@ -252,35 +259,80 @@ class ArrayView
    */
   const std::size_t n_elements;
 
-  friend class ArrayView<const ElementType>;
+  friend class ArrayView<const ElementType, MemorySpaceType>;
 };
 
 
 
 //---------------------------------------------------------------------------
 
 
+namespace internal
+{
+  namespace ArrayViewHelper
+  {
+    template <typename MemorySpaceType>
+    inline bool
+    is_in_correct_memory_space(const void *const ptr)
+    {
+#ifndef DEAL_II_COMPILER_CUDA_AWARE
+      (void)ptr;
+      static_assert(std::is_same<MemorySpaceType, MemorySpace::Host>::value,
+                    "If the compiler doesn't understand CUDA code, "
+                    "the only possible memory space is 'MemorySpace::Host'!");
+      return true;
+#else
+      cudaPointerAttributes attributes;
+      const cudaError_t cuda_error = cudaPointerGetAttributes(&attributes, ptr);
+      if (cuda_error != cudaErrorInvalidValue)
+        {
+          AssertCuda(cuda_error);
+          if (std::is_same<MemorySpaceType, MemorySpace::Host>::value)
+            return attributes.memoryType == cudaMemoryTypeHost;
+          else
+            return attributes.memoryType == cudaMemoryTypeDevice;
```

Comment:
yes, that's the sort of thing I had in mind!

Reply:
I don't think that's a good idea. We got very strange bugs when using `#ifdef` inside a function. Now, you potentially have the function define twice with different implementations and you never know which one you will get. I think that this change should be reverted.

Reply:
As long as the `static_assert` is in place, we can't define it twice for `MemorySpace::CUDA` so I think it is safe.

Reply:
But you still have two different implementations for `MemorySpace::Host` with one of them always returning `true`.